### PR TITLE
[scheduler] Add SchedulerConfig and validation tests

### PR DESCRIPTION
**Pull request by CaseyHillers**
Part of go/flutter-build-test-infra

This change is a no-op to Cocoon, and only adds models necessary for future changes.

* Create `SchedulerConfig` and `Target` protos
* Validate scheduler config from a yaml is valid
  * Ensure all dependencies exist
  * No duplicate targets
  * No cycles (the config should only create a tree)
* Add `bin/validate_scheduler_config.dart` for running tests against external configs

# Issues

https://github.com/flutter/flutter/issues/76140

### CODE REVIEW

File: app_dart/lib/src/model/proto/internal/scheduler.proto [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+message SchedulerConfig {
+    // Targets to run from this config.
+    repeated Target targets = 1;
+    // Git branches to run these targets against.
+    repeated string enabled_branches = 2;
+}
+
+message Target {
+    // Unique, human readable identifier.
+    optional string name = 1;
```
Comment (keyonghan) [inline app_dart/lib/src/model/proto/internal/scheduler.proto:16]: shall the name be required? is it builder name, task name, or any other?

File: app_dart/lib/src/model/proto/internal/scheduler.proto [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+syntax = "proto2";
+
+message SchedulerConfig {
+    // Targets to run from this config.
+    repeated Target targets = 1;
+    // Git branches to run these targets against.
+    repeated string enabled_branches = 2;
+}
+
+message Target {
```
Comment (keyonghan) [inline app_dart/lib/src/model/proto/internal/scheduler.proto:14]: Do we need a flaky flag somewhere?

File: app_dart/lib/src/service/scheduler.dart [hunk 3]
```diff
@@ -256,3 +257,57 @@ class Scheduler {
...
+  // Construct [targetGraph]. With a one scan approach, cycles in the graph
+  // cannot exist as it only works forward.
+  for (final Target target in schedulerConfig.targets) {
+    if (targetGraph.containsKey(target.name)) {
+      exceptions.add('ERROR: ${target.name} already exists in graph');
+    } else {
+      targetGraph[target.name] = <Target>[];
+      // Add edges
+      if (target.dependencies.isNotEmpty) {
+        if (target.dependencies.length != 1) {
```
Comment (keyonghan) [inline app_dart/lib/src/service/scheduler.dart:290]: why is `dependencies` defined as repeated then?

File: app_dart/lib/src/service/scheduler.dart [hunk 4]
```diff
@@ -256,3 +257,57 @@ class Scheduler {
...
+          exceptions
+              .add('ERROR: ${target.name} has multiple dependencies which is not supported. Use only one dependency');
+        } else {
+          if (target.dependencies.first == target.name) {
+            exceptions.add('ERROR: ${target.name} cannot depend on itself');
+          } else if (targetGraph.containsKey(target.dependencies.first)) {
+            targetGraph[target.dependencies.first].add(target);
```
Comment (keyonghan) [inline app_dart/lib/src/service/scheduler.dart:297]: do we need to maintain the child list of a target?

File: app_dart/test/service/scheduler_test.dart [hunk 5]
```diff
@@ -250,6 +252,135 @@ void main() {
...
+  - master
+targets:
+  - name: A
+    dependencies:
+      - B
+  - name: B
+    dependencies:
+      - A
+      ''') as YamlMap;
+      expect(() => loadSchedulerConfig(configWithCycle), throwsA(isA<FormatException>()));
```
Comment (keyonghan) [inline app_dart/test/service/scheduler_test.dart:343]: can we verify the exception summary for different scenarios?

Comment (CaseyHillers) [reply app_dart/test/service/scheduler_test.dart]: Added!

File: app_dart/lib/src/service/scheduler.dart (hunk 4 above)
Comment (CaseyHillers) [reply app_dart/lib/src/service/scheduler.dart:297]: This graph is specifically for validating configs. For future scheduling, we'll want to use a hash map for tracking what builds should be triggered when. So when `$target` notifies of completion, we call `triggers[$target]` to quickly lookup the next set.
```
trigger['web_engine_linux'] = <Target>['web_engine_linux_chrome', 'web_engine_linux_firefox', ...];
```

File: app_dart/lib/src/service/scheduler.dart (hunk 3 above)
Comment (CaseyHillers) [reply app_dart/lib/src/service/scheduler.dart:290]: This is future functionality we'd like to support, but it adds a lot of complexity to the initial scheduler. It's easier to support if if this is already a list

File: app_dart/lib/src/model/proto/internal/scheduler.proto (hunk 2 above)
Comment (CaseyHillers) [reply app_dart/lib/src/model/proto/internal/scheduler.proto:14]: From feedback on go/flutter-build-test-infra, we're renaming it as bringup

File: app_dart/lib/src/model/proto/internal/scheduler.proto (hunk 1 above)
Comment (CaseyHillers) [reply app_dart/lib/src/model/proto/internal/scheduler.proto:16]: Protos should never use required field as it causes issues in the future if you deprecate the field ([go/protodosdonts](http://go/protodosdonts))

**Review by keyonghan — APPROVED** [review]
LGTM

**Comment by CaseyHillers** [comment]
\cc @christopherfujino

**Pull request merged**